File: scripts/core-canonical-identity-oracle.py
#!/usr/bin/env python3
"""CORE-ID-001 explicit cross-language canonical lexical identity oracle."""

from __future__ import annotations

import argparse
from pathlib import Path

PROFILE_ID = "symthaea.core.canonical-lexical-id.v1"

EDGE_WHITESPACE_CODEPOINTS = frozenset(
    {
        0x0009,
        0x000A,
        0x000B,
        0x000C,
        0x000D,
        0x0020,
        0x0085,
        0x00A0,
        0x1680,
        0x2000,
        0x2001,
        0x2002,
        0x2003,
        0x2004,
        0x2005,
        0x2006,
        0x2007,
        0x2008,
        0x2009,
        0x200A,
        0x2028,
        0x2029,
        0x202F,
        0x205F,
        0x3000,
    }
)


class CanonicalIdentityError(ValueError):
    """Input is not admitted by the explicit CORE-ID-001 lexical profile."""


def is_explicit_edge_whitespace(ch: str) -> bool:
    return ord(ch) in EDGE_WHITESPACE_CODEPOINTS


def is_forbidden_control(ch: str) -> bool:
    cp = ord(ch)
    return cp < 0x20 or cp == 0x7F


def admit_utf8_bytes(data: bytes, max_utf8_bytes: int) -> str:
    if not isinstance(data, bytes):
        raise CanonicalIdentityError("bytes required")
    if not isinstance(max_utf8_bytes, int) or isinstance(max_utf8_bytes, bool):
        raise CanonicalIdentityError("positive integer byte limit required")
    if max_utf8_bytes <= 0:
        raise CanonicalIdentityError("positive integer byte limit required")
    if not data:
        raise CanonicalIdentityError("identifier required")
    if len(data) > max_utf8_bytes:
        raise CanonicalIdentityError("identifier exceeds UTF-8 byte limit")
    try:
        value = data.decode("utf-8", "strict")
    except UnicodeDecodeError as exc:
        raise CanonicalIdentityError("valid UTF-8 required") from exc

    if is_explicit_edge_whitespace(value[0]) or is_explicit_edge_whitespace(value[-1]):
        raise CanonicalIdentityError("declared edge whitespace is non-canonical")
    if any(is_forbidden_control(ch) for ch in value):
        raise CanonicalIdentityError("ASCII C0/DEL control is non-canonical")
    return value


def admit_text(value: str, max_utf8_bytes: int) -> bytes:
    if not isinstance(value, str):
        raise CanonicalIdentityError("text required")
    try:
        encoded = value.encode("utf-8", "strict")
    except UnicodeEncodeError as exc:
        raise CanonicalIdentityError("valid UTF-8 required") from exc
    admit_utf8_bytes(encoded, max_utf8_bytes)
    return encoded


def _accepts_bytes(data: bytes, max_utf8_bytes: int) -> bool:
    try:
        admit_utf8_bytes(data, max_utf8_bytes)
    except CanonicalIdentityError:
        return False
    return True


def _legacy_pie_accepts_text(value: str, max_utf8_bytes: int = 4096) -> bool:
    """Qualified PIE-ID-001 Python 3.13.5 acceptance relation."""
    try:
        encoded = value.encode("utf-8", "strict")
    except UnicodeEncodeError:
        return False
    if not encoded or len(encoded) > max_utf8_bytes:
        return False
    if value != value.strip():
        return False
    if any(ord(ch) < 0x20 or ord(ch) == 0x7F for ch in value):
        return False
    return True


def _explicit_accepts_text(value: str, max_utf8_bytes: int = 4096) -> bool:
    try:
        admit_text(value, max_utf8_bytes)
    except CanonicalIdentityError:
        return False
    return True


def prove_legacy_pie_edge_equivalence() -> None:
    """Exhaustively compare every Unicode code point at both identifier edges."""
    for cp in range(0x110000):
        ch = chr(cp)
        for value in (ch + "x", "x" + ch):
            if _legacy_pie_accepts_text(value) != _explicit_accepts_text(value):
                raise AssertionError(f"legacy/explicit edge mismatch at U+{cp:04X}")


def _decode_hex(value: str) -> bytes:
    try:
        return bytes.fromhex(value)
    except ValueError as exc:
        raise CanonicalIdentityError("invalid vector hex") from exc


def run_vectors(path: Path) -> tuple[str, ...]:
    out: list[str] = []
    for lineno, raw in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
        if not raw or raw.startswith("#"):
            continue
        parts = raw.split("\t")
        if len(parts) != 4:
            raise CanonicalIdentityError(f"vector line {lineno}: expected 4 fields")
        name, max_bytes_text, hex_text, expected = parts
        max_bytes = int(max_bytes_text)
        data = _decode_hex(hex_text)
        actual = "accept" if _accepts_bytes(data, max_bytes) else "reject"
        if expected not in {"accept", "reject"}:
            raise CanonicalIdentityError(f"vector line {lineno}: invalid expected status")
        if actual != expected:
            raise AssertionError(f"vector {name}: expected {expected}, got {actual}")
        out.append(f"{name}\t{actual}")
    if not out:
        raise AssertionError("vector corpus must not be empty")
    return tuple(out)


def self_test() -> None:
    assert PROFILE_ID == "symthaea.core.canonical-lexical-id.v1"
    assert len(EDGE_WHITESPACE_CODEPOINTS) == 25

    assert len(admit_text("a" * 4096, 4096)) == 4096
    assert not _explicit_accepts_text("a" * 4097, 4096)
    assert len(admit_text("é" * 2048, 4096)) == 4096
    assert not _explicit_accepts_text("é" * 2049, 4096)

    assert len(admit_text("a" * 256, 256)) == 256
    assert not _explicit_accepts_text("a" * 257, 256)

    assert _explicit_accepts_text("p\u00a01")
    assert _explicit_accepts_text("\u200Bx")
    assert _explicit_accepts_text("\uFEFFx")
    assert not _explicit_accepts_text("\u00A0x")
    assert not _explicit_accepts_text("x\u3000")

    assert admit_text("Process-A", 4096) != admit_text("process-a", 4096)
    assert admit_text("é", 4096) != admit_text("e\u0301", 4096)

    prove_legacy_pie_edge_equivalence()


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--vectors", type=Path)
    parser.add_argument("--self-test", action="store_true")
    args = parser.parse_args()

    if args.self_test:
        self_test()
        print("ok")
        return 0
    if args.vectors is not None:
        for line in run_vectors(args.vectors):
            print(line)
        return 0
    parser.error("choose --vectors or --self-test")


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/core-canonical-identity-oracle.rs
// CORE-ID-001 explicit cross-language canonical lexical identity oracle.

use std::env;
use std::fs;
use std::path::Path;

const PROFILE_ID: &str = "symthaea.core.canonical-lexical-id.v1";

fn is_explicit_edge_whitespace(ch: char) -> bool {
    matches!(
        ch as u32,
        0x0009..=0x000D
            | 0x0020
            | 0x0085
            | 0x00A0
            | 0x1680
            | 0x2000..=0x200A
            | 0x2028
            | 0x2029
            | 0x202F
            | 0x205F
            | 0x3000
    )
}

fn is_forbidden_control(ch: char) -> bool {
    let cp = ch as u32;
    cp < 0x20 || cp == 0x7F
}

fn admit_utf8_bytes(data: &[u8], max_utf8_bytes: usize) -> bool {
    if max_utf8_bytes == 0 || data.is_empty() || data.len() > max_utf8_bytes {
        return false;
    }
    let value = match std::str::from_utf8(data) {
        Ok(value) => value,
        Err(_) => return false,
    };
    let first = match value.chars().next() {
        Some(ch) => ch,
        None => return false,
    };
    let last = match value.chars().next_back() {
        Some(ch) => ch,
        None => return false,
    };
    if is_explicit_edge_whitespace(first) || is_explicit_edge_whitespace(last) {
        return false;
    }
    !value.chars().any(is_forbidden_control)
}

fn hex_nibble(byte: u8) -> Option<u8> {
    match byte {
        b'0'..=b'9' => Some(byte - b'0'),
        b'a'..=b'f' => Some(byte - b'a' + 10),
        b'A'..=b'F' => Some(byte - b'A' + 10),
        _ => None,
    }
}

fn decode_hex(text: &str) -> Result<Vec<u8>, String> {
    let bytes = text.as_bytes();
    if !bytes.len().is_multiple_of(2) {
        return Err("hex input must have even length".to_string());
    }
    let mut out = Vec::with_capacity(bytes.len() / 2);
    for pair in bytes.chunks_exact(2) {
        let hi = hex_nibble(pair[0]).ok_or_else(|| "invalid hex digit".to_string())?;
        let lo = hex_nibble(pair[1]).ok_or_else(|| "invalid hex digit".to_string())?;
        out.push((hi << 4) | lo);
    }
    Ok(out)
}

fn run_vectors(path: &Path) -> Result<Vec<String>, String> {
    let text = fs::read_to_string(path).map_err(|error| error.to_string())?;
    let mut out = Vec::new();
    for (index, raw) in text.lines().enumerate() {
        if raw.is_empty() || raw.starts_with('#') {
            continue;
        }
        let parts: Vec<&str> = raw.split('\t').collect();
        if parts.len() != 4 {
            return Err(format!("vector line {}: expected 4 fields", index + 1));
        }
        let name = parts[0];
        let max_utf8_bytes = parts[1]
            .parse::<usize>()
            .map_err(|_| format!("vector {name}: invalid byte limit"))?;
        let data = decode_hex(parts[2])?;
        let actual = if admit_utf8_bytes(&data, max_utf8_bytes) {
            "accept"
        } else {
            "reject"
        };
        let expected = parts[3];
        if expected != "accept" && expected != "reject" {
            return Err(format!("vector {name}: invalid expected status"));
        }
        if actual != expected {
            return Err(format!(
                "vector {name}: expected {expected}, got {actual}"
            ));
        }
        out.push(format!("{name}\t{actual}"));
    }
    if out.is_empty() {
        return Err("vector corpus must not be empty".to_string());
    }
    Ok(out)
}

fn require(condition: bool, message: &str) -> Result<(), String> {
    if condition {
        Ok(())
    } else {
        Err(message.to_string())
    }
}

fn self_test() -> Result<(), String> {
    require(
        PROFILE_ID == "symthaea.core.canonical-lexical-id.v1",
        "profile id mismatch",
    )?;

    let ascii_4096 = vec![b'a'; 4096];
    let ascii_4097 = vec![b'a'; 4097];
    require(admit_utf8_bytes(&ascii_4096, 4096), "4096-byte boundary")?;
    require(
        !admit_utf8_bytes(&ascii_4097, 4096),
        "4097-byte overflow",
    )?;

    let ascii_256 = vec![b'a'; 256];
    let ascii_257 = vec![b'a'; 257];
    require(admit_utf8_bytes(&ascii_256, 256), "256-byte boundary")?;
    require(!admit_utf8_bytes(&ascii_257, 256), "257-byte overflow")?;

    let multibyte_4096 = "é".repeat(2048);
    let multibyte_4098 = "é".repeat(2049);
    require(
        multibyte_4096.len() == 4096 && admit_utf8_bytes(multibyte_4096.as_bytes(), 4096),
        "multibyte exact boundary",
    )?;
    require(
        !admit_utf8_bytes(multibyte_4098.as_bytes(), 4096),
        "multibyte overflow",
    )?;

    require(
        admit_utf8_bytes("p\u{00A0}1".as_bytes(), 4096),
        "interior NBSP must remain admitted",
    )?;
    require(
        admit_utf8_bytes("\u{200B}x".as_bytes(), 4096),
        "zero-width space is not V1 edge whitespace",
    )?;
    require(
        admit_utf8_bytes("\u{FEFF}x".as_bytes(), 4096),
        "BOM is not V1 edge whitespace",
    )?;
    require(
        !admit_utf8_bytes("\u{00A0}x".as_bytes(), 4096),
        "leading NBSP must be rejected",
    )?;
    require(
        !admit_utf8_bytes("x\u{3000}".as_bytes(), 4096),
        "trailing ideographic space must be rejected",
    )?;

    require(
        "Process-A".as_bytes() != "process-a".as_bytes(),
        "case must remain byte-distinct",
    )?;
    require(
        "é".as_bytes() != "e\u{0301}".as_bytes(),
        "normalization must not be implicit",
    )?;

    Ok(())
}

fn main() -> Result<(), String> {
    let args: Vec<String> = env::args().collect();
    match args.as_slice() {
        [_, flag] if flag == "--self-test" => {
            self_test()?;
            println!("ok");
            Ok(())
        }
        [_, flag, path] if flag == "--vectors" => {
            self_test()?;
            for line in run_vectors(Path::new(path))? {
                println!("{line}");
            }
            Ok(())
        }
        _ => Err("choose --vectors <path> or --self-test".to_string()),
    }
}
